src/Agda2Hs/Compile/Type.hs

```diff
@@ -155,42 +107,67 @@ compileType t = do
          | Just semantics <- isSpecialType f -> setCurrentRange f $ semantics es
          | Just args <- allApplyElims es ->
              ifJustM (isUnboxRecord f) (\_ -> compileUnboxType f args) $
-             ifM (isTransparentFunction f) (compileTransparentType args) $
+             ifM (isTransparentFunction f) (compileTransparentType (defType def) args) $
              ifM (isInlinedFunction f) (compileInlineType f es) $ do
-               vs <- compileTypeArgs args
+               vs <- compileTypeArgs (defType def) args
                f <- compileQName f
                return $ tApp (Hs.TyCon () f) vs
          | otherwise -> fail
 
     Var x es | Just args <- allApplyElims es -> do
-      vs <- compileTypeArgs args
+      xi <- lookupBV x
+      unless (usableModality xi) $ genericDocError
+        =<< text "Cannot use erased variable" <+> prettyTCM (var x)
+        <+> text "in Haskell type"
+      vs <- compileTypeArgs (snd $ unDom xi) args
       x  <- hsName <$> compileDBVar x
       return $ tApp (Hs.TyVar () x) vs
 
     Sort s -> return (Hs.TyStar ())
 
-    Lam argInfo restAbs
-      | not (keepArg argInfo) -> underAbstraction_ restAbs compileType
+    -- This case is only reachable for erased lambdas
```

Comment:
Is this true?
I need to test locally but say, if the type is `(f : (Type -> Type) -> Type) (x : f (\y -> Nat)) -> Nat` or smthg, wouldn't the lambda be allowed without the check?

Reply:
You are right, this is actually a real problem.
```agda
foo : (f : (Set → Set) → Set) (x : f (λ y → Nat)) (y : f List) → Nat
foo f x y = 42

{-# COMPILE AGDA2HS foo #-}
```
is now compiled to 
```haskell
foo :: f Natural -> f [] -> Natural
foo x y = 42
```
where `f` is applied to two arguments of a different kind. I'll add the check again that we don't compile non-erased lambdas.